src/structures/paging/recursive.rs

```diff
@@ -109,25 +110,26 @@ impl<'a> RecursivePageTable<'a> {
         allocator: &mut A,
     ) -> Result<MapperFlush<S>, MapToError>
     where
-        A: FnMut() -> Option<PhysFrame>,
+        A: PhysMemAllocator,
         S: PageSize,
         Self: Mapper<S>,
     {
         let page = Page::containing_address(VirtAddr::new(frame.start_address().as_u64()));
         self.map_to(page, frame, flags, allocator)
     }
 
-    fn create_next_table<A>(
+    fn create_next_table<A, S>(
         entry: &mut PageTableEntry,
         allocator: &mut A,
     ) -> Result<bool, MapToError>
     where
-        A: FnMut() -> Option<PhysFrame>,
+        A: PhysMemAllocator,
+        S: PageSize,
     {
         use structures::paging::PageTableFlags as Flags;
 
         if entry.is_unused() {
-            if let Some(frame) = allocator() {
+            if let Some(frame) = allocator.alloc::<S>() {
```

Comment:
A page table is always 4KiB large, so it does not make sense to call this function with a generic `S`.